crates/router/src/core/user.rs

```diff
@@ -1937,6 +1957,40 @@ pub async fn check_two_factor_auth_status(
     ))
 }
 
+pub async fn check_two_factor_auth_status_with_attempts(
+    state: SessionState,
+    user_token: auth::UserIdFromAuth,
+) -> UserResponse<user_api::TwoFactorStatus> {
+    let user_from_db: domain::UserFromStorage = state
+        .global_store
+        .find_user_by_id(&user_token.user_id)
+        .await
+        .change_context(UserErrors::InternalServerError)?
+        .into();
+    if user_from_db.get_totp_status() == TotpStatus::NotSet {
+        return Ok(ApplicationResponse::Json(user_api::TwoFactorStatus {
+            status: None,
+        }));
+    };
+
+    Ok(ApplicationResponse::Json(user_api::TwoFactorStatus {
+        status: Some(user_api::TwoFactorAuthStatusResponseWithAttempts {
+            totp: user_api::TwoFactorAuthAttempts {
+                is_completed: tfa_utils::check_totp_in_redis(&state, &user_token.user_id).await?,
+                remaining_attempts: consts::user::TOTP_MAX_ATTEMPTS
+                    - tfa_utils::check_totp_attempts_in_redis(&state, &user_token.user_id).await?,
+            },
+            recovery_code: user_api::TwoFactorAuthAttempts {
+                is_completed: tfa_utils::check_recovery_code_in_redis(&state, &user_token.user_id)
+                    .await?,
+                remaining_attempts: consts::user::RECOVERY_CODE_MAX_ATTEMPTS
+                    - tfa_utils::check_recovery_code_attempts_in_redis(&state, &user_token.user_id)
+                        .await?,
```

Comment:
NIT: It's better to split this into 2 variables like `totp` and `recovery_code` and construct this.